Added basic text about enqueue_file_kwargs (#6273)
sunpy/sunpy: docs/dev_guide/contents/extending_fido.rst

The search method should in this case, generate a ``baseurl`` dependant on the values of these attrs, and then call ``super().search`` or `~sunpy.net.scraper.Scraper` for each ``baseurl`` generated.
For an example of a complex modification of the ``search()`` method see the implementation of `.SUVIClient.search`.

Customizing the Downloader
--------------------------

There is no method for a client creator to override the `parfive.Downloader` that is used to fetch the files.
This is because all downloads made by a single call to ``Fido.fetch`` share one instance of `parfive.Downloader`.
However, it is possible to pass keywords :meth:`parfive.Downloader.enqueue_file`, which is important if there is a need to customise the requests to a remote server, such as setting custom HTTP headers.

> One example is the `sunpy.net.dataretriever.sources.noaa.SRSClient`:

Reviewer: ```suggestion
This is done by setting the ``enqueue_file_kwargs`` attribute of the client class.
One example from the `sunpy.net.dataretriever.sources.noaa.SRSClient` is:
```

.. code-block:: python

>     # Server does not support the normal aioftp passive command.
>     enqueue_file_kwargs = {"passive_commands": ["pasv"]}

Reviewer: ```suggestion
    class SRSClient(GenericClient):
        ...
        # Server does not support the normal aioftp passive command.
        enqueue_file_kwargs = {"passive_commands": ["pasv"]}
        ...
```

> These keywords are passed to each call to :meth:`parfive.Downloader.enqueue_file`, so they will affect all files that are downloaded, by your client.

Reviewer: ```suggestion
These keywords are passed to each call to :meth:`parfive.Downloader.enqueue_file`, so they will affect all files that are added for download by your client.
```

The search method should in this case, generate a ``baseurl`` dependant on the values of these attrs, and then call ``super().search`` or `~sunpy.net.scraper.Scraper` for each ``baseurl`` generated.
For an example of a complex modification of the ``search()`` method see the implementation of `.SUVIClient.search`.

Customizing the Downloader
--------------------------

> There is no method for a client creator to override the `parfive.Downloader` that is used to fetch the final files.

Reviewer: ```suggestion
There is no method for a client creator to override the `parfive.Downloader` that is used to fetch the files.
```

> This is because we allow a user to setup a custom `parfive.Downloader` by passing it to the ``downloader`` argument of :meth:`Fido.fetch`.

Reviewer: While technically true, the more important statement is that all downloads from all clients share a single `Downloader` instance, be that the one passed by the user or the one created in `Fido.fetch`.

```suggestion
This is because all downloads made by a single call to :meth:`Fido.fetch` share one instance of `parfive.Downloader`.
```

> However, it is possible to pass keywords :meth:`parfive.Downloader.enqueue_file`, which is important if need to work around a limitation on the remote server.

Reviewer 2: ```suggestion
However, it is possible to pass keywords :meth:`parfive.Downloader.enqueue_file`, which is important if there is a need to work around a limitation on the remote server.
```

> However, it is possible to pass keywords :meth:`parfive.Downloader.enqueue_file`, which is important if there is a need to work around a limitation on the remote server.

Reviewer: ```suggestion
However, it is possible to pass keywords :meth:`parfive.Downloader.enqueue_file`, which is important if there is a need to customise the requests to a remote server, such as setting custom HTTP headers.
```

One example is the `sunpy.net.dataretriever.sources.noaa.SRSClient`:

.. code_block:: python

    # Server does not support the normal aioftp passive command.
    enqueue_file_kwargs = {"passive_commands": ["pasv"]}

> These keywords are passed to each call to :meth:`parfive.Downloader.enqueue_file`, so they will affect all files that are downloaded.

Reviewer: ```suggestion
These keywords are passed to each call to :meth:`parfive.Downloader.enqueue_file`, so they will affect all files that are downloaded, by your client.
```

It does this by converting the input query to a set of queries which are ORed, but are complete queries.
This means the list of queries is an **OR** of **ANDs** (technically called `disjuntive normal form <https://en.wikipedia.org/wiki/Disjunctive_normal_form>`__).

Each query in the list of ORs contains all the information about that query so for example if the user provided a query like

.. code-block:: python

    a.Time("2020/02/02", "2020/02/03") & (a.Instrument("AIA") | a.Instrument("HMI"))

> it would be passed to the client as::

Author: ```suggestion
it would be passed to the client as
```

.. code-block:: python

>     (a.Time("2020/02/02", "2020/02/03") & a.Instrument("HMI")) | (a.Time("2020/02/02", "2020/02/03") & a.Instrument("AIA"))

Reviewer: This code block isn't rendering properly.

Reviewer 3: There's a `::` at the end of the previous line so it's all being interpreted as a code block.

                results.append(self._make_search(query_parameters))

            return QueryResponseTable(results, client=self)

        def _make_filename(path, row, resp, url):
            # Define a fallback filename based on the information in the search results
            name = f"row['ID'].fits"

            if resp:
                cdheader = resp.headers.get("Content-Disposition", None)
                if cdheader:
                _, params = cgi.parse_header(cdheader)
                name = params.get('filename', "")

            return path.format(file=name, **row.response_block_map)

        @convert_row_to_table
        def fetch(self, query_results, *, path, downloader, **kwargs):
            for row in query_results:
                filepath = partial(self._make_filename, path, row)

                url = f"https://sfsi.sunpy.org/download/{row['ID']}"
                downloader.enqueue_file(url, filename=filepath)

        @classmethod
        def _can_handle_query(cls, *query):
            query_attrs = set(type(x) for x in query)
            supported_attrs = {a.Time, a.Level}
            return supported_attrs.issuperset(query_attrs)

> Customizing the Downloader

Reviewer: This is in the wrong section, it only applies to the simple GenericClient part.
